uvicorn/protocols/websockets/websockets_impl.py

```diff
@@ -278,43 +282,83 @@ async def asgi_send(self, message: "ASGISendEvent") -> None:
         message_type = message["type"]
 
         if not self.handshake_started_event.is_set():
-            if message_type == "websocket.accept":
-                message = cast("WebSocketAcceptEvent", message)
-                self.logger.info(
-                    '%s - "WebSocket %s" [accepted]',
-                    self.scope["client"],
-                    get_path_with_query_string(self.scope),
-                )
-                self.initial_response = None
-                self.accepted_subprotocol = cast(
-                    Optional[Subprotocol], message.get("subprotocol")
-                )
-                if "headers" in message:
-                    self.extra_headers.extend(
-                        # ASGI spec requires bytes
-                        # But for compatibility we need to convert it to strings
+            if self.response_body is None:
```

Comment:
Can't we remove this conditional and add `self.handshake_started_event.set()` on the `websocket.http.response.start`?

Reply:
We need to somehow distinguish between the case when handshake is started with `websocket.http.response.start` as opposed to the others, because it sets us in a state where the only allowable messages are `websocket.http.response.body`. Otherwise, we cannot perform state validation.

We also cannot set the event until the body is ready, because we `await` this event (line 214) and expect any response to be ready when it is set.  This event really should be renamed, I could call it `handshake_done_event` if you agree.

I could however avoid the extra `if` clause (and indentation) and keep handling of all these handshake messages on the same level, but it would complicate the message validation logic.

Reply:
Do you think we could benefit from another event to be used on line 214?

Reply:
if too much work, don't worry